src/main/java/org/assertj/core/internal/Comparables.java

```diff
@@ -153,12 +154,12 @@ public String toString() {
    *           {@code org.junit.ComparisonFailure} instead if JUnit is in the classpath and the expected and actual
    *           values are not equal.
    */
-  public <T extends Comparable<? super T>> void assertEqualByComparison(AssertionInfo info, T actual, T expected) {
+  @SuppressWarnings("unchecked")
+  public <T> void assertEqualByComparison(AssertionInfo info, Comparable<? super T> actual, Comparable<? super T> expected) {
     assertNotNull(info, actual);
     // we don't delegate to comparisonStrategy, as this assertion makes it clear it relies on Comparable
-    if (actual.compareTo(expected) == 0)
-      return;
-    throw failures.failure(info, shouldBeEqual(actual, expected, info.representation()));
+    // TODO are we sure that any instance of Comparable<? super T> is always a T?
+    if (actual.compareTo((T) expected) != 0) throw failures.failure(info, shouldBeEqual(actual, expected, info.representation()));
```

Comment:
@scordio as I changed the generic signature `expected` is now a `Comparable<T>` but not a `T` and thus `actual.compareTo(expected)` does not compile. I "fixed" that with a cast but I'm not 100% sure we won't get a class cast exception (maybe not as generics are erased at runtime).
I could not find a type that would be a `Comparable<T>` but not a `T`, do you think that cast can blow up?

Reply:
I was able to put together something weird enough that can throw a `ClassCastException`:

```java
static class BaseClass implements Comparable<Subclass> {
  @Override
  public int compareTo(Subclass o) {
    return 0;
  }
}

static class Subclass extends BaseClass {
}

@Test
void test() {
  // GIVEN
  Subclass s1 = new Subclass();
  BaseClass b1 = new BaseClass();
  // WHEN/THEN
  assertThatComparable(s1).isEqualByComparingTo(b1);
}
```

yields:

```
java.lang.ClassCastException: class org.assertj.core.api.Assertions_assertThatComparable_Test$BaseClass cannot be cast to class org.assertj.core.api.Assertions_assertThatComparable_Test$Subclass
```

Same result with the changes in #2587.

Reply:
Good example, I think it is very unlikely to encounter a type like that because one could argue that defining a base class in term of its subclasses is not a great design.

Reply:
I agree, it's a pretty rare use. I'll give it another try to see if we can prevent it.

Reply:
I couldn't find a way to prevent:

```
assertThatComparable(s1).isEqualByComparingTo(b1);
```

at compile time without breaking other use cases, knowing that:

```
s1.compareTo(b1);
```

doesn't compile.

This is because we want to provide a `Comparable` as input parameter for the assertion method, while `compareTo` directly accepts the value type.

We could make it work but we would take out the option of providing `Comparable<T>` as assertion parameter, like in the following examples:

```java
Comparable<Name> name1 = new Name("ABC"); // object under assertion can be Comparable

Name name2 = new Name("ABC");
Comparable<Name> comparableName2 = name2;

assertThatComparable(name1).isEqualByComparingTo(name2); // compiles
assertThatComparable(name1).isEqualByComparingTo(comparableName2); // does not compile
```

I tend to favor type-safety at the price of not supporting the case with `comparableName2`.

What do you think @joel-costigliola ?

Reply:
I agree, I think `assertThatComparable(name1).isEqualByComparingTo(name2);` is more commonly seen than `assertThatComparable(name1).isEqualByComparingTo(comparableName2);`